plugin/acl/acl.go

```diff
@@ -0,0 +1,99 @@
+package acl
+
+import (
+	"context"
+	"net"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/metrics"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+
+	"github.com/infobloxopen/go-trees/iptree"
+	"github.com/miekg/dns"
+)
+
+var log = clog.NewWithPlugin("acl")
+
+type acl struct {
+	Next plugin.Handler
+
+	Rules []Rule
+}
+
+// Rule defines a list of Zones and some ACL policies which will be
+// enforced on them.
+type Rule struct {
+	Zones    []string
+	Policies []Policy
+}
+
+// Policy defines the ACL policy for DNS queries.
+// A policy performs the specified action (block/allow) on all DNS queries
+// matched by source IP or QTYPE.
+type Policy struct {
+	action int
+	qtypes map[uint16]struct{}
+	filter *iptree.Tree
+}
+
+const (
+	// Allow allows authorized queries to recurse.
+	Allow = iota
+	// Block blocks unauthorized queries towards protected DNS zones.
+	Block
+)
+
+func (a acl) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	for _, rule := range a.Rules {
+		// check zone.
+		zone := plugin.Zones(rule.Zones).Matches(state.Name())
+		if zone == "" {
+			continue
+		}
+
+		if shouldBlock(rule.Policies, w, r) {
+			m := new(dns.Msg)
+			m.SetRcode(r, dns.RcodeRefused)
+			w.WriteMsg(m)
+			RequestBlockCount.WithLabelValues(metrics.WithServer(ctx), zone).Inc()
+			return dns.RcodeSuccess, nil
+		}
```

Comment:
should we add here 
```
else {
  break
}
```
If `shouldBlock()` returned `false` this means we found appropriate rule and the action was allow. Why we continue iterating over remaining rules?

Reply:
Well that depends on how you treat the relationship between different rules. Is it `OR` or `AND`? I think `AND` make more sense here because it supports a stricter control.

Besides, `shouldBlock` returning false only means that for this rule, the request should not be blocked. For example, not matched by any ACL policies. So using `break` here might not work.

Reply:
So, in conflicting cases `block` action will have higher priority than `allow`?
BTW, is it possible/legitimate to create several ACLs with the same or intersecting set of zones?

Reply:
> in conflicting cases block action will have higher priority than allow?

Logically yes.

> is it possible/legitimate to create several ACLs with the same or intersecting set of zones?

No, I don't think so. It does not make any sense to have conflicts in ACL rules. Ideally we should detect this and warn user, but it's really hard to do this.

Reply:
Although I'm not sure whether there is a case where users have to introduce conflicts in ACLs, using a fine-grained subdomain should handle most of the cases:

```
example.org {
  acl a.example.org {
    block ...
    // ...
  }
  acl b.example.org c.example.org {
    // ...
  }
}
```

Reply:
That's why I think iterating over remaining rules is just wasting time. In fine-grained config there should be no conflicts.
In case of conflicts, with my proposal the first encountered action will win. In your implementation `block` wins.

Reply:
Yes u r right. What about this:
``` golang
isBlock, matched := shouldBlock(...)
if isBlock {
  // return dns.RcodeRefused.
}
if matched {
  // return with recurse
}
// continue checking with other rules.
```

Reply:
FYI: https://github.com/coredns/coredns/pull/3103/commits/d09b3f65d8e959dff224ff4ef2dfa11ea83ac46e

Reply:
Oh, I see, you want to distinguish between allow and no match.

Alternatively, you can define
```
type action int
const (
  actionNone = iota
  actionAllow
  actionBlock
)
...
func matchWithPolicies(policies []Policy, w dns.ResponseWriter, r *dns.Msg) action {
   ...
}
```
and return `actionNone` if no policy match

Reply:
Thanks for suggestion. Now updated.